File: SOLUTIONS/LEETCODE/sudoku/validSudoku/validSudoku.js
function isValidSudoku(board) {
  // use sets to track seen numbers for rows, columns and subgrids
  const rows = Array.from({length: 9}, () => new Set())
  const columns = Array.from({length: 9}, () => new Set())
  const subgrids = Array.from({ length: 9 }, () => new Set())
  
  for (let r = 0; r < 9; r++){
    for (let c = 0; c < 9; c++){
      let value = board[r][c]
      if (value === ".") continue   // skip empty cells
      
      // check row
      if (rows[r].has(value)) { return false }
      rows[r].add(value)

      // check columns
      if (columns[c].has(value)) { return false }
      columns[c].add(value)

      // check subgrid
      // formula is like that of diagonal trick calculation -> map cell to subgrid index
      let subgridIndex = Math.floor(r / 3) * 3 + Math.floor(c / 3)
      
      if (subgrids[subgridIndex].has(value)) { return false }
      subgrids[subgridIndex].add(value)
    }
  }

  // because we are concerned about the truthfulness of validity, we;
  return true
}

File: SOLUTIONS/LEETCODE/sudoku/validSudoku/validSudoku.py
def isValidSudoku(board):
  # Use sets to track seen numbers for rows, cols, and subgrids
  rows = [set() for _ in range(9)]
  columns = [set() for _ in range(9)]
  subgrids = [set() for _ in range(9)]
  
  # loop through rows and columns and subgrid
  for r in range(9):
    for c in range(9):
      value = board[r][c]
  
      if value == ".":
        continue   # skip empty cells
      
      if value in rows: return False
      rows[r].add(value)
      
      if value in columns: return False
      columns[c].add(value)
      
      # Formula: map each cell(r,c) -> a 0-8 index for its 3x3 subgrids
      subgridIndex = (r//3) * 3 + (c//3)
      if value in subgrids[subgridIndex]: return False
      subgrids[subgridIndex].add(value)
       
  return True
      
      